python/ray/_private/client_mode_hook.py

```diff
@@ -91,13 +95,23 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
-def client_mode_should_convert():
-    # This is for testing with RAY_CLIENT_MODE.
-    # When RAY_CLIENT_MODE=1, it means that for all the tests
-    # will run with client mode.
-    # is_client_mode_enabled will be set to be off when client is off
+def client_mode_should_convert(*, auto_init: bool = True):
+    """Determines if functions should be converted to client mode & if
+    Ray should be auto-initialized.
+
+    NOTE: `auto_init` must happen before we branch into regular ray or client
+    code because the initialization may result in either mode.
+    """
+    if auto_init:
```

Comment:
I believe `auto_init` is an existing feature. Where is `auto_init` happening before / elsewhere?

Reply:
It is done at the location of `is_connected` checks (for `state.py` and `worker.py`):
https://github.com/ray-project/ray/blob/fd13bac9b3fc2e7142065c759f2c9fc1c753e912/python/ray/worker.py#L225
https://github.com/ray-project/ray/blob/fd13bac9b3fc2e7142065c759f2c9fc1c753e912/python/ray/state.py#L53

Reply:
I have seen people confused by auto init in some cases, e.g. starting a new Ray cluster instead of connecting to the local node. We may want to optimize the heuristics in future. Is it possible to consolidate the auto init logic, or add reminders about where to update the logic?